### PR TITLE
add cluster autoscaler options to hosted clusted

### PR DESCRIPTION
**Pull request by elmiko**
This change adds the initial cluster autoscaler field options to the
HostedCluster type. It also adds tests to confirm the fields are
populating the deployment for the autoscaler. There is also a small
update to the api doc to ensure parity with the code.

### CODE REVIEW

**Comment by elmiko** [comment]
this is still wip, i'd like to add a few more tests but i wanted to take advantage of the ci here.

**Comment by elmiko** [comment]
fixed an error and added another test

**Comment by elmiko** [comment]
fixed up a couple errors in the value printing for int stuff

**Comment by elmiko** [comment]
i ran the `make api` on this, but it didn't produce any new output. just curious if i missed something?

**Comment by elmiko** [comment]
added the proper api files and fixed the verify error

File: api/v1alpha1/hostedcluster_types.go [hunk 1]
```diff
@@ -142,6 +145,29 @@ type Release struct {
...
 
+// TODO maybe we have profiles for scaling behaviors
+type ClusterAutoscaling struct {
+	// Maximum number of nodes in all node groups.
+	// Cluster autoscaler will not grow the cluster beyond this number.
+	// +kubebuilder:validation:Minimum=0
+	MaxNodesTotal *int32 `json:"maxNodesTotal,omitempty"`
+
+	// Gives pods graceful termination time before scaling down
+	// default: 600 seconds
```
Comment (enxebre) [inline api/v1alpha1/hostedcluster_types.go:156]: can we do any :validation here?

File: api/v1alpha1/hostedcluster_types.go [hunk 2]
```diff
@@ -32,6 +32,9 @@ type HostedClusterSpec struct {
 	// Networking contains network-specific settings for this cluster
 	Networking ClusterNetworking `json:"networking"`
 
+	// Autoscaling for compute nodes only, does not cover control plane
+	Autoscaling ClusterAutoscaling `json:"autoscaling"`
```
Comment (enxebre) [inline api/v1alpha1/hostedcluster_types.go:36]: should this `omitempty`?

Comment (elmiko) [reply api/v1alpha1/hostedcluster_types.go:36]: i wasn't sure, but given that all the options inside the struct are omitempty it probably makes sense here as well.

File: api/v1alpha1/hostedcluster_types.go (hunk 1 above)
Comment (elmiko) [reply api/v1alpha1/hostedcluster_types.go:156]: not sure i follow, could you explain a little more please?

Comment (enxebre) [reply api/v1alpha1/hostedcluster_types.go:156]: Is there a value for this fields that would make the cluster autoscaler fail? If so can we structural schema validate that here so the bad input does not make it through?

Comment (elmiko) [reply api/v1alpha1/hostedcluster_types.go:156]: ah! thanks for clarifying that. i'll have to check, but i think the only invalid values would be negative numbers. i will double check though.

**Comment by elmiko** [comment]
updated
* added omitempty to autoscaler entry in hostedcluster
* added minimum 0 validation to max grace period timeout field

File: hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go [hunk 3]
```diff
@@ -275,3 +277,90 @@ func TestComputeHostedClusterAvailability(t *testing.T) {
 		})
 	}
 }
+
+func listIncludesString(list []string, target string) bool {
```
Comment (ironcladlou) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:281]: This function can be deleted and replaced by using `sets.String` which gives you nice set semantics

Comment (elmiko) [reply hypershift-operator/controllers/hostedcluster/hostedcluster_controller_test.go:281]: thanks! i haven't used that before, i'll check it out.

**Comment by enxebre** [comment]
thanks @elmiko!
Since we don't have e2e validation for autoscaling, have you tested this?
/test e2e-aws
/lgtm
/hold

Feel free to hold cancel after testing this.

**Comment by elmiko** [comment]
> Since we don't have e2e validation for autoscaling, have you tested this?

just tested locally to ensure that the args were coming out correctly. i haven't tried it on a live cluster yet, i can do that early next week.

**Comment by elmiko** [comment]
i tested this out manually, and everything seems to be in place. the default (empty) spec produces the current behavior, and the options are all editable on the hostedcluster resource. for example

hostedcluster snippet
```
spec:                                                                                                                                                                   
  autoscaling:                                                                                                                                                          
    maxNodeProvisionTime: 45m                                                                                                                                           
    maxNodesTotal: 100                                                                                                                                                  
    maxPodGracePeriod: 1000                                                                                                                                             
    podPriorityThreshold: -5
```

deployment output
```
    spec:                                                                                                                                                               
      containers:                  
      - args:                                                                       
        - --cloud-provider=clusterapi
        - --node-group-auto-discovery=clusterapi:namespace=$(MY_NAMESPACE)          
        - --kubeconfig=/mnt/kubeconfig/target-kubeconfig                            
        - --clusterapi-cloud-config-authoritative
        - --alsologtostderr
        - --v=4
        - --max-nodes-total=100
        - --max-graceful-termination-sec=1000
        - --max-node-provision-time=45m
        - --expendable-pods-priority-cutoff=-5
```

/hold cancel

**Pull request merged**